File: 0495. Teemo Attacking/0495.java
// time complexity: O(n*duration)
// space complexity: O(1)
class Solution {
    public int findPoisonedDuration(int[] timeSeries, int duration) {
        // timeSeries array is non-decreasing -> for ex: [1,1,1,2,5,6]

        int total = 0;
        for(int i = 0; i < timeSeries.length; i++) {
            // skip duplicate attack times
            if( i < timeSeries.length - 1 && ,
            timeSeries[i] == timeSeries[i + 1])
            continue;

            // simulate poison effect for each second
            for(int j = 0; j < duration; j++) {
                total++;
                // if the next attack happens before poison ends, stop early
                if( i < timeSeries.length - 1 &&
                        timeSeries[i] + j + 1 == timeSeries[i + 1])
                    break;
            }
        }
        return total;
    }
}

File: 0495. Teemo Attacking/0495.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    def findPoisonedDuration(self, timeSeries, duration):
        """
        :type timeSeries: List[int]
        :type duration: int
        :rtype: int
        """
        total = 0
        # iterate through attacks except the last one
        for i in range(len(timeSeries) - 1):
            # minimum(duration, the gap to the next attack)
            total += min(duration, timeSeries[i+1] - timeSeries[i])
        # add poison duration for the last attack
        total += duration
        return total